README.md
**countryguess** looks up country information by country codes or name matching.
It tries to be lean (but not mean) and fast: All dependencies are in the Python
Standard Library and country data is loaded lazily on demand.

Code: [Codeberg](https://codeberg.org/plotski/countryguess)  
Package: [PyPI](https://pypi.org/project/countryguess)

### Usage

`guess_country()` uses the default country data that is already packaged.

```python
>>> from countryguess import guess_country
>>> guess_country("britain")
{
    'name_short': 'United Kingdom',
    'name_official': 'United Kingdom of Great Britain and Northern Ireland',
    'iso2': 'GB',
    'iso3': 'GBR',
    ...
}
>>> guess_country("no such country")
None
>>> guess_country("no such country", default="Oh, well.")
'Oh, well.'
>>> guess_country("PoRtUgAl", attribute="iso2")
'PT'
>>> guess_country("TW", attribute="name_official")  # 2-letter code lookup
'Republic of China'
>>> guess_country("tWn", attribute="name_short")    # 3-letter code lookup
'Taiwan'
```

Matching by regular expression can be extended by mapping
[ISO2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) codes to
[`re.Pattern`](https://docs.python.org/3/library/re.html#re.compile) objects.

```python
>>> regex_map = {
...     "MN": re.compile(r'^mongol\s+uls$', flags=re.IGNORECASE),
...     "JP": re.compile(r'^ni(?:pp|h)on', flags=re.IGNORECASE),
... }
>>> guess_country("Mongol Uls", attribute="name_short", regex_map=regex_map)
'Mongolia'
>>> guess_country("Nippon", attribute="name_short", regex_map=regex_map)
'Japan'
```

You can also create a `CountryData` instance yourself to provide your own
country data in a JSON file.

```python
>>> from countryguess import CountryData
>>> countries = CountryData("path/to/countries.json")
>>> countries["vIeTnAm"]
{'name_short': 'Vietnam', ...}
>>> countries["vn"]
{'name_short': 'Vietnam', ...}
>>> countries["asdf"]
KeyError: 'asdf'
>>> countries.get("asdf")
None
>>> countries.get("kuwait")
{'name_short': 'Kuwait', ...}
```

On `CountryData` instances, every key in the JSON data is accessible as a
method.

```python
>>> countries.name_official("portugal")
'Portuguese Republic'
>>> countries.continent("vanuatu")
'Oceania'
```

### Country Lookup

Countries are identified by name, 2-letter code
([ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)) or
3-letter code
([ISO 3166-1 alpha-3](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3)). All
identifiers are matched case-insensitively.

Names are matched with regular expressions that are stored in the JSON data. If
that fails, fuzzy matching against ``name_short`` and ``name_official`` is done
with [difflib](https://docs.python.org/3/library/difflib.html).

### Country Data

Country information is read from a JSON file. One is shipped with the package,
but you can also provide your own to the `CountryData` class as described
above. The information in the default file was gratefully extracted from
[country-converter](https://pypi.org/project/country-converter/). (Many thanks!)

The country data file must contain a list of JSON objects. Each object
represents a country that must contain at least the following keys:

- `name_short`
- `name_official`
- `iso2`
- `iso3`
- `regex`

#### Packaged Classification Schemes

The following classification schemes are available in the included country data.

<!-- CLASSIFICATION_SCHEMES (see fetch_data_from_country_converter.py) -->
1.  ISO2 (ISO 3166-1 alpha-2)
2.  ISO3 (ISO 3166-1 alpha-3)
3.  ISO - numeric (ISO 3166-1 numeric)
4.  UN numeric code (M.49 - follows to a large extend ISO-numeric)
5.  A standard or short name
6.  The "official" name
7.  Continent
8.  UN region
9.  [EXIOBASE](http://exiobase.eu/) 1 classification
10. [EXIOBASE](http://exiobase.eu/) 2 classification
11. [EXIOBASE](http://exiobase.eu/) 3 classification
12. [WIOD](http://www.wiod.org/home) classification
13. [Eora](http://www.worldmrio.com/)
14. [OECD](http://www.oecd.org/about/membersandpartners/list-oecd-member-countries.htm)
    membership (per year)
15. [MESSAGE](http://www.iiasa.ac.at/web/home/research/researchPrograms/Energy/MESSAGE-model-regions.en.html)
    11-region classification
16. [IMAGE](https://models.pbl.nl/image/index.php/Welcome_to_IMAGE_3.0_Documentation)
17. [REMIND](https://www.pik-potsdam.de/en/institute/departments/transformation-pathways/models/remind)
18. [UN](http://www.un.org/en/members/) membership (per year)
19. [EU](https://ec.europa.eu/eurostat/statistics-explained/index.php/Glossary:EU_enlargements)
    membership (including EU12, EU15, EU25, EU27, EU27_2007, EU28)
20. [EEA](https://ec.europa.eu/eurostat/statistics-explained/index.php/Glossary:European_Economic_Area_(EEA))
    membership
21. [Schengen](https://en.wikipedia.org/wiki/Schengen_Area) region
22. [Cecilia](https://cecilia2050.eu/system/files/De%20Koning%20et%20al.%20%282014%29_Scenarios%20for%202050_0.pdf)
    2050 classification
23. [APEC](https://en.wikipedia.org/wiki/Asia-Pacific_Economic_Cooperation)
24. [BRIC](https://en.wikipedia.org/wiki/BRIC)
25. [BASIC](https://en.wikipedia.org/wiki/BASIC_countries)
26. [CIS](https://en.wikipedia.org/wiki/Commonwealth_of_Independent_States)
    (as by 2019, excl. Turkmenistan)
27. [G7](https://en.wikipedia.org/wiki/Group_of_Seven)
28. [G20](https://en.wikipedia.org/wiki/G20) (listing all EU member
    states as individual members)
29. [FAOcode](http://www.fao.org/faostat/en/#definitions) (numeric)
30. [GBDcode](http://ghdx.healthdata.org/) (numeric - Global Burden of
    Disease country codes)
31. [IEA](https://www.iea.org/countries) (World Energy Balances 2021)
32. [DACcode](https://www.oecd.org/dac/financing-sustainable-development/development-finance-standards/dacandcrscodelists.htm)
    (numeric - OECD Development Assistance Committee)
33. [ccTLD](https://en.wikipedia.org/wiki/Country_code_top-level_domain) - country code top-level domains
34. [GWcode](https://www.tandfonline.com/doi/abs/10.1080/03050629908434958) - Gledisch & Ward numerical codes as published in https://www.andybeger.com/states/articles/statelists.html
<!-- CLASSIFICATION_SCHEMES -->

### Command Line Interface

**countryguess** comes with a simple CLI with the same name. It takes one or two
arguments:

```sh
$ countryguess oman
{
    "name_short": "Oman",
    "name_official": "Sultanate of Oman",
    ...
}
$ countryguess 'puerto ricco' name_official
Puerto Rico
```

### Contributing

All kinds of bug reports, feature requests and suggestions are welcome!


NEWS
0.?.?
  * Faster lookup with custom regex_map argument


0.3.0
  * guess_country() and CountryData.get() get a new regex_map argument
    (see README)
  * Updated metadata from upstream country converter
    https://github.com/IndEcol/country_converter


0.2.1
  * Updated metadata from upstream country converter
    https://github.com/IndEcol/country_converter


0.2.0
  * Added country attributes: cctld, gwcode


0.1.0
  * Initial release


tests/countrydata_test.py
import copy
import io
import re
import sys
from unittest.mock import Mock, call

import pytest

from countryguess import __project_name__, _countrydata


def test_lazy_load_countries(mocker):
    self = Mock(
        _countries=None,
        _filepath='path/to/countrydata',
    )
    mocks = Mock()
    mocks.attach_mock(self.func, 'func')
    mocks.attach_mock(self._load_countries, '_load_countries')
    self._load_countries.return_value = 'mock countries'

    wrapped = _countrydata._lazy_load_countries(self.func)
    for _ in range(6):
        return_value = wrapped(self, 1, two='3')

    assert return_value is self.func.return_value
    assert mocks.mock_calls == [
        call._load_countries(),
        call.func(self, 1, two='3'),
        call.func(self, 1, two='3'),
        call.func(self, 1, two='3'),
        call.func(self, 1, two='3'),
        call.func(self, 1, two='3'),
        call.func(self, 1, two='3'),
    ]
    assert self._countries is mocks._load_countries.return_value


@pytest.mark.parametrize(
    argnames='property_name',
    argvalues=(
        'countries',
        'codes_iso2',
        'codes_iso3',
        'names_official',
        'names_short',
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_lazy_loaded_property(property_name):
    countrydata = _countrydata.CountryData()
    assert countrydata._countries is None
    getattr(countrydata, property_name)
    assert isinstance(countrydata._countries, list)


@pytest.mark.parametrize(
    argnames='method_name, args, kwargs',
    argvalues=(
        ('_find_country', ('antarctica',), {}),
        ('__getattr__', ('iso2',), {}),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_lazy_loaded_method(method_name, args, kwargs):
    countrydata = _countrydata.CountryData()
    assert countrydata._countries is None
    method = getattr(countrydata, method_name)
    method(*args, **kwargs)
    assert isinstance(countrydata._countries, list)


@pytest.mark.parametrize(
    argnames='filepath, filecontent, exp_countries',
    argvalues=(
        (
            'custom/countries.json',
            '[{"name_short": "Customland", "regex": "^custom$"}]',
            [{'name_short': 'Customland', 'regex': re.compile('^custom$', flags=re.IGNORECASE)}],
        ),
        (
            None,
            '[{"name_short": "Kingdom of Default", "regex": "^default$"}]',
            [{'name_short': 'Kingdom of Default', 'regex': re.compile('^default$', flags=re.IGNORECASE)}],
        ),
    ),
    ids=lambda v: repr(v),
)
def test_load_countries(filepath, filecontent, exp_countries, tmp_path, mocker):
    if sys.version_info <= (3, 9, 0):
        # TODO: Remove this when Python 3.9 is no longer supported
        open_text_mock = mocker.patch('importlib.resources.open_text', return_value=(
            io.StringIO('[{"name_short": "Kingdom of Default", "regex": "^default$"}]')
        ))

        def assert_expectations():
            assert open_text_mock.call_args_list == [call(__project_name__, '_countrydata.json')]

    else:
        # Python 3.13 removes importlib.resources.open_text()
        files_mock = mocker.patch('importlib.resources.files')
        package_path_mock = files_mock.return_value
        joinpath_mock = package_path_mock.joinpath
        file_path_mock = joinpath_mock.return_value
        open_mock = file_path_mock.open
        open_mock.return_value = (
            io.StringIO('[{"name_short": "Kingdom of Default", "regex": "^default$"}]')
        )

        def assert_expectations():
            assert files_mock.mock_calls == [
                call(__project_name__),
                call().joinpath('_countrydata.json'),
                call().joinpath().open('r', encoding='utf8'),
            ]

    if filepath:
        filepath = tmp_path / filepath
        filepath.parent.mkdir(parents=True, exist_ok=True)
        filepath.write_text(filecontent)

    countrydata = _countrydata.CountryData(filepath)
    return_value = countrydata._load_countries()
    assert return_value == exp_countries

    if filepath is None:
        assert_expectations()


def test_CountryData_property_countries():
    countrydata = _countrydata.CountryData()
    orig_countries = copy.deepcopy(countrydata.countries)
    countries = countrydata.countries
    assert countries == countrydata.countries
    del countries[0]
    countries[0]['name_official'] = 'Foo Country'
    assert countries != countrydata.countries
    assert orig_countries == countrydata.countries


property_test_data = '''[
{"iso3": "ABC", "iso2": "AB", "name_short": "Foo", "name_official": "The Foo", "regex": "irrelevant"},
{"iso3": "DEF", "iso2": "DE", "name_short": "Bar", "name_official": "The Bar", "regex": "irrelevant"},
{"iso3": "GHI", "iso2": "GH", "name_short": "Baz", "name_official": "The Baz", "regex": "irrelevant"}
]'''

@pytest.mark.parametrize(
    argnames='property_name, exp_value',
    argvalues=(
        ('codes_iso3', ('ABC', 'DEF', 'GHI')),
        ('codes_iso2', ('AB', 'DE', 'GH')),
        ('names_short', ('Foo', 'Bar', 'Baz')),
        ('names_official', ('The Foo', 'The Bar', 'The Baz')),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_property(property_name, exp_value, tmp_path):
    filepath = tmp_path / 'countrydata.json'
    filepath.write_text(property_test_data)
    countrydata = _countrydata.CountryData(filepath)

    value = getattr(countrydata, property_name)
    assert value == exp_value

    # Return value is cached
    for _ in range(6):
        assert value is getattr(countrydata, property_name)


@pytest.fixture
def find_country_countries():
    return [
        {
            'iso2': 'AB',
            'iso3': 'ABC',
            'name_short': 'Foolala',
            'name_official': 'Fooland',
            'regex': re.compile('^fo+lala$', flags=re.IGNORECASE),
        },
        {
            'iso2': 'DE',
            'iso3': 'DEF',
            'name_short': 'Baristan',
            'name_official': 'Republic of Baristan',
            'regex': re.compile('^bah?ristan$', flags=re.IGNORECASE),
        },
        {
            'iso2': 'GH',
            'iso3': 'GHI',
            'name_short': 'Bazvia',
            'name_official': 'Bazvian Kingdom',
            'regex': re.compile('^ba[zs]via$', flags=re.IGNORECASE),
        },
    ]

@pytest.mark.parametrize(
    argnames='string, exp_info_index',
    argvalues=(
        ('ab', 0),
        ('De', 1),
        ('GH', 2),
        ('xy', None),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_find_country_by_iso2(string, exp_info_index, find_country_countries):
    countrydata = _countrydata.CountryData()
    countrydata._countries = find_country_countries
    info = countrydata._find_country(string)
    if exp_info_index is None:
        assert info is None
    else:
        info_index = countrydata._countries.index(info)
        assert info_index == exp_info_index

@pytest.mark.parametrize(
    argnames='string, exp_info_index',
    argvalues=(
        ('abc', 0),
        ('DeF', 1),
        ('GHI', 2),
        ('xyz', None),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_find_country_by_iso3(string, exp_info_index, find_country_countries):
    countrydata = _countrydata.CountryData()
    countrydata._countries = find_country_countries
    info = countrydata._find_country(string)
    if exp_info_index is None:
        assert info is None
    else:
        info_index = countrydata._countries.index(info)
        assert info_index == exp_info_index

@pytest.mark.parametrize(
    argnames='string, regex_map, exp_info_index',
    argvalues=(
        (
            'aaaaaabbbbbb',
            {
                'ab': re.compile(r'^a+b+$'),
                'de': re.compile(r'^dee?EEE?fff?$'),
                'gh': re.compile(r'^.*$'),
            },
            0,
        ),
        (
            'deEEEff',
            {
                'ab': re.compile(r'^a+b+$'),
                'de': re.compile(r'^dee?EEE?fff?$'),
                'gh': re.compile(r'^.*$'),
            },
            1,
        ),
        (
            'anything',
            {
                'ab': re.compile(r'^a+b+$'),
                'de': re.compile(r'^dee?EEE?fff?$'),
                'gh': re.compile(r'^[a-z]+$'),
            },
            2,
        ),
        (
            '123',
            {
                'ab': re.compile(r'^a+b+$'),
                'de': re.compile(r'^dee?EEE?fff?$'),
                'gh': re.compile(r'^[a-z]+$'),
            },
            None,
        ),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_find_country_by_custom_regex(string, regex_map, exp_info_index, find_country_countries, mocker):
    countrydata = _countrydata.CountryData()
    mocker.patch.object(countrydata, '_validate_regex_map')
    countrydata._countries = find_country_countries
    info = countrydata._find_country(string, regex_map=regex_map)
    if exp_info_index is None:
        assert info is None
    else:
        info_index = countrydata._countries.index(info)
        assert info_index == exp_info_index

    if exp_info_index is not None:
        assert countrydata._validate_regex_map.call_args_list == []
    else:
        assert countrydata._validate_regex_map.call_args_list == [call(regex_map)]

@pytest.mark.parametrize(
    argnames='string, exp_info_index',
    argvalues=(
        ('FOOlala', 0),
        ('foooolala', 0),
        ('fOOOooolala', 0),
        ('baristan', 1),
        ('Bahristan', 1),
        ('BARistan', 1),
        ('BAHRistan', 1),
        ('Bazvia', 2),
        ('baZvia', 2),
        ('Basvia', 2),
        ('BASvia', 2),
        ('this should not match any country', None),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_find_country_by_hardcoded_regex(string, exp_info_index, find_country_countries):
    countrydata = _countrydata.CountryData()
    countrydata._countries = find_country_countries
    info = countrydata._find_country(string)
    if exp_info_index is None:
        assert info is None
    else:
        info_index = countrydata._countries.index(info)
        assert info_index == exp_info_index

@pytest.mark.parametrize(
    argnames='string, exp_info_index',
    argvalues=(
        ('Foolalu', 0),
        ('Foolalulala', None),
        ('Brepublic bof Baristan', 1),
        ('Brepublic bof Baristanianism', None),
        ('Basvian Kingdork', 2),
        ('Basvian Queendom', None),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_find_country_by_fuzzy_match(string, exp_info_index, find_country_countries):
    countrydata = _countrydata.CountryData()
    countrydata._countries = find_country_countries
    info = countrydata._find_country(string)
    if exp_info_index is None:
        assert info is None
    else:
        info_index = countrydata._countries.index(info)
        assert info_index == exp_info_index


@pytest.mark.parametrize(
    argnames='regex_map, exp_exception',
    argvalues=(
        (['a', 'b'], RuntimeError("Not a dict-like object: ['a', 'b']")),
        ({'a': 'b'}, RuntimeError("Not a ISO 3166-1 alpha-2 country code: 'a'")),
        ({'TV': 'b'}, RuntimeError("Not a regular expression (see re.compile()): 'b'")),
        ({'TV': re.compile('toovaloodldoo')}, None),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_validate_regex_map(regex_map, exp_exception):
    countrydata = _countrydata.CountryData()
    if isinstance(exp_exception, Exception):
        with pytest.raises(type(exp_exception), match=rf'^{re.escape(str(exp_exception))}$'):
            countrydata._validate_regex_map(regex_map)
    else:
        countrydata._validate_regex_map(regex_map)


@pytest.mark.parametrize(
    argnames='country, default, find_country, exp_return_value',
    argvalues=(
        ('foo', None, Mock(return_value={'iso2': 'FO'}), {'iso2': 'FO'}),
        ('foo', 'default', Mock(return_value={'iso2': 'FO'}), {'iso2': 'FO'}),
        ('foo', None, Mock(return_value=None), None),
        ('foo', 'default', Mock(return_value=None), 'default'),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_get(country, default, find_country, exp_return_value, mocker):
    regex_map = Mock()
    mocker.patch('countryguess._countrydata.CountryData._find_country', find_country)
    countrydata = _countrydata.CountryData()
    return_value = countrydata.get(country, default=default, regex_map=regex_map)
    assert return_value == exp_return_value
    assert countrydata._find_country.call_args_list == [
        call(country, regex_map=regex_map)
    ]


@pytest.mark.parametrize(
    argnames='country, get, exp_result',
    argvalues=(
        ('foo', Mock(return_value={'iso2': 'FO'}), {'iso2': 'FO'}),
        ('foo', Mock(return_value=None), KeyError('foo')),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_getitem(country, get, exp_result, mocker):
    mocker.patch('countryguess._countrydata.CountryData.get', get)
    countrydata = _countrydata.CountryData()
    if isinstance(exp_result, Exception):
        with pytest.raises(type(exp_result), match=rf'^{re.escape(str(exp_result))}$'):
            countrydata[country]
    else:
        return_value = countrydata[country]
        assert return_value == exp_result


@pytest.mark.parametrize(
    argnames='attribute, country, exp_result',
    argvalues=(
        ('iso2', 'foo', 'FO'),
        ('iso3', 'foo', 'FOO'),
        ('iso4', 'foo', AttributeError('iso4')),
    ),
    ids=lambda v: repr(v),
)
def test_CountryData_getattr(attribute, country, exp_result, mocker):
    countrydata = _countrydata.CountryData()
    countrydata._countries = [
        {
            'iso2': 'FO',
            'iso3': 'FOO',
        },
    ]
    if isinstance(exp_result, Exception):
        with pytest.raises(type(exp_result), match=rf'^{re.escape(str(exp_result))}$'):
            getattr(countrydata, attribute)
    else:
        return_value = getattr(countrydata, attribute)(country)
        assert return_value == exp_result


tests/guess_country_test.py
import re
from unittest.mock import call

import pytest

from countryguess import _guess_country


@pytest.fixture(autouse=True)
def remove_CountryData_instances():
    try:
        yield
    finally:
        _guess_country._countrydata = None


def test_guess_country_loads_countrydata_on_demand(mocker):
    CountryData_mock = mocker.patch('countryguess._guess_country.CountryData')
    assert _guess_country._countrydata is None
    for _ in range(12):
        _guess_country.guess_country('foo')
    assert _guess_country._countrydata is CountryData_mock.return_value
    CountryData_mock.call_args_list == [call()]


@pytest.mark.parametrize(
    argnames='info, attribute, default, exp_result',
    argvalues=(
        ({'iso2': 'AB', 'iso3': 'ABC'}, None, None, {'iso2': 'AB', 'iso3': 'ABC'}),
        ({'iso2': 'AB', 'iso3': 'ABC'}, None, 'whatever', {'iso2': 'AB', 'iso3': 'ABC'}),
        ({'iso2': 'AB', 'iso3': 'ABC'}, 'iso2', None, 'AB'),
        ({'iso2': 'AB', 'iso3': 'ABC'}, 'ISO2', None, 'AB'),
        ({'iso2': 'AB', 'iso3': 'ABC'}, 'iso3', None, 'ABC'),
        ({'iso2': 'AB', 'iso3': 'ABC'}, 'iso3', 'whatever', 'ABC'),
        ({'iso2': 'AB', 'iso3': 'ABC'}, 'iso4', None, AttributeError('iso4')),
        (None, None, 'whatever', 'whatever'),
        (None, 'iso3', 'whatever', 'whatever'),
    ),
    ids=lambda v: repr(v),
)
def test_guess_country_return_value(info, attribute, default, exp_result, mocker):
    CountryData_mock = mocker.patch('countryguess._guess_country.CountryData')
    CountryData_mock.return_value.get.return_value = info

    if isinstance(exp_result, Exception):
        with pytest.raises(type(exp_result), match=rf'^{re.escape(str(exp_result))}$'):
            _guess_country.guess_country('foo', attribute=attribute, default=default)
    else:
        return_value = _guess_country.guess_country('foo', attribute=attribute, default=default)
        assert return_value == exp_result

    assert CountryData_mock.return_value.get.call_args_list == [call('foo', regex_map=None)]


fetch_data_from_country_converter.py
#!/usr/bin/env python3

import csv
import io
import json
import os
import re
import subprocess
import urllib.request

upstream_url = 'https://github.com/konstantinstadler/country_converter/'
cwd = os.path.dirname(__file__)

country_data_url = f'{upstream_url}/raw/master/country_converter/country_data.tsv'
country_data_file = os.path.join(cwd, 'countryguess/_countrydata.json')

readme_url = f'{upstream_url}/raw/master/README.md'
readme_file = os.path.join(cwd, 'README.md')


def show_diff(path):
    subprocess.run(['git', '-C', cwd, '--no-pager', 'diff', '--word-diff', path])


def fetch_country_data():
    patches = {
        # https://github.com/konstantinstadler/country_converter/issues/92
        # https://github.com/konstantinstadler/country_converter/commit/6a06051e915e76a9ffc29fe0c08808cf403df99f
        # https://github.com/konstantinstadler/country_converter/blob/master/country_converter/country_data.tsv#L242
        'United Kingdom of Great Britain and Northern Ireland': {
            'iso2': 'GB',
        },
    }

    with urllib.request.urlopen(country_data_url) as request:
        # Read TSV data
        response_string = request.read().decode('utf-8')
        data = csv.DictReader(io.StringIO(response_string), dialect=csv.excel_tab)
        country_list = [
            {
                k.lower(): v
                for k, v in country.items()
            }
            for country in data
        ]

        # Apply patches
        for country in country_list:
            if country['name_official'] in patches:
                for k, v in patches[country['name_official']].items():
                    country[k] = v

        # Overwrite packaged file
        json_data = json.dumps(country_list, indent=2)
        with open(country_data_file, 'w') as f:
            f.write(json_data + '\n')

        show_diff(country_data_file)


def fetch_classifications_from_README():

    def get_classification_schemes():
        with urllib.request.urlopen(readme_url) as request:
            text = request.read().decode('utf-8')
        # with open('../country_converter/README.md') as request:
        #     text = request.read()
            match = re.search(
                r'## Classification schemes.*?(1\..*?)(?:\s*\n\s*){2,}',
                text,
                flags=re.DOTALL | re.MULTILINE,
            )
            if match:
                return match.group(1)
            else:
                raise RuntimeError('Failed to extract classification schemes')

    def replace_classification_schemes():
        readme = open(readme_file, 'r').read()
        schemes_regex = re.compile(
            r'<!-- CLASSIFICATION_SCHEMES.*CLASSIFICATION_SCHEMES -->',
            flags=re.DOTALL | re.MULTILINE,
        )
        schemes_template = (
            '<!-- CLASSIFICATION_SCHEMES (see fetch_data_from_country_converter.py) -->\n'
            '{schemes}\n'
            '<!-- CLASSIFICATION_SCHEMES -->'
        )

        if not re.search(schemes_regex, readme):
            raise RuntimeError('Failed to find classification schemes in README.md')

        readme = re.sub(
            schemes_regex,
            schemes_template.format(schemes=get_classification_schemes()),
            readme,
        )
        with open(readme_file, 'w') as f:
            f.write(readme)

    replace_classification_schemes()

    show_diff(readme_file)


fetch_country_data()
fetch_classifications_from_README()


countryguess/_countrydata.py
import collections
import difflib
import functools
import importlib.resources
import json
import re
import sys

from . import __project_name__


def _lazy_load_countries(func):
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        # Read country data from file unless we've already done that
        if not self._countries:
            self._countries = self._load_countries()

        # Call wrapped function transparently
        return func(self, *args, **kwargs)

    return wrapper


class CountryData:
    def __init__(self, filepath=None):
        self._filepath = filepath
        self._countries = None

    def _load_countries(self):
        if self._filepath is not None:
            stream = open(self._filepath, 'r')
        else:
            if sys.version_info <= (3, 9, 0):
                # TODO: Remove this when Python 3.9 is no longer supported
                stream = importlib.resources.open_text(__project_name__, '_countrydata.json')
            else:
                # Python 3.13 removes importlib.resources.open_text()
                package_path = importlib.resources.files(__project_name__)
                file_path = package_path.joinpath('_countrydata.json')
                stream = file_path.open('r', encoding='utf8')

        country_list = json.load(stream)

        for info in country_list:
            info['regex'] = re.compile(info['regex'], flags=re.IGNORECASE)

        return country_list

    @property
    @_lazy_load_countries
    def countries(self):
        """
        :class:`list` of country :class:`dict` objects

        This is the same data that was read from the provided country data file.
        """
        return [country.copy() for country in self._countries]

    @functools.cached_property
    @_lazy_load_countries
    def codes_iso2(self):
        """Sequence of ISO 3166-1 alpha-2 country codes"""
        return tuple(country['iso2'] for country in self._countries)

    @functools.cached_property
    @_lazy_load_countries
    def codes_iso3(self):
        """Sequence of ISO 3166-1 alpha-3 country codes"""
        return tuple(country['iso3'] for country in self._countries)

    @functools.cached_property
    @_lazy_load_countries
    def names_official(self):
        """Sequence of official country names"""
        return tuple(country['name_official'] for country in self._countries)

    @functools.cached_property
    @_lazy_load_countries
    def names_short(self):
        """Sequence of colloqial country names"""
        return tuple(country['name_short'] for country in self._countries)

    @_lazy_load_countries
    def _find_country(self, string, regex_map=None):
        # ISO 3166-1 alpha-2
        if len(string) == 2:
            info = self._find_country_by_code(string, self.codes_iso2)
            if info:
                return info

        # ISO 3166-1 alpha-3
        if len(string) == 3:
            info = self._find_country_by_code(string, self.codes_iso3)
            if info:
                return info

        # Custom regular expressions
        if regex_map:
            self._validate_regex_map(regex_map)
            for iso2, regex in regex_map.items():
                if regex.search(string):
                    return self._find_country_by_code(iso2, self.codes_iso2)

        # Hardcoded regular expressions
        for info in self._countries:
            if info['regex'].search(string):
                return info

        # Fuzzy country name
        for names in (self.names_official, self.names_short):
            matches = difflib.get_close_matches(string, names, n=1, cutoff=0.8)
            if matches:
                name = matches[0]
                index = names.index(name)
                return self._countries[index]

    def _validate_regex_map(self, regex_map):
        if not isinstance(regex_map, collections.abc.Mapping):
            raise RuntimeError(f'Not a dict-like object: {regex_map!r}')

        for iso2, regex in regex_map.items():
            if iso2 not in self.codes_iso2:
                raise RuntimeError(f'Not a ISO 3166-1 alpha-2 country code: {iso2!r}')

            elif not isinstance(regex, re.Pattern):
                raise RuntimeError(f'Not a regular expression (see re.compile()): {regex!r}')

    def _find_country_by_code(self, code, codes):
        try:
            index = codes.index(code.upper())
        except ValueError:
            pass
        else:
            return self._countries[index]

    def get(self, country, default=None, regex_map=None):
        """
        Return country data as :class:`dict`

        :param str country: Country name, 2-letter code or 3-letter code

            This is case-insensitive.

            If a country name is provided, it is first matched against the
            regular expressions. If that fails,
            :func:`difflib.get_close_matches` is used for fuzzy matching.

        :param default: Default return value if `country` is not found
        :param dict regex_map: Map ISO 3166-1 alpha-2 country codes
            (:class:`str`) to regular expressions (:class:`re.Pattern`, see
            :func:`re.compile`)
        """
        info = self._find_country(country, regex_map=regex_map)
        if info:
            return info
        else:
            return default

    def __getitem__(self, country):
        info = self.get(country)
        if info:
            return info
        else:
            raise KeyError(country)

    @_lazy_load_countries
    def __getattr__(self, attribute):
        # Raise exception now, not when get_attribute() is called. This means
        # accessing `countrydata.iso4` raises AttributeError as it should.
        try:
            self._countries[0][attribute]
        except KeyError:
            raise AttributeError(attribute)

        def get_attribute(country):
            info = self.get(country)
            return info[attribute]

        return get_attribute
